maven-embedder/src/main/java/org/apache/maven/cli/props/MavenPropertiesLoader.java

```diff
@@ -0,0 +1,176 @@
+                (k, v) -> sp.put(k.toString(), escape ? InterpolationHelper.escape(v.toString()) : v.toString()));
+        substitute(sp, callback, INCLUDES_PROPERTY);
+        loadIncludes(INCLUDES_PROPERTY, path, sp);
+        substitute(sp, callback);
+        sp.forEach(properties::setProperty);
+    }
+
+    public static void substitute(MavenProperties props, Function<String, String> callback) {
+        for (Enumeration<?> e = props.propertyNames(); e.hasMoreElements(); ) {
+            String name = (String) e.nextElement();
+            String value = props.getProperty(name);
+            if (value == null) {
+                value = callback.apply(name);
+            }
+            if (name.startsWith(OVERRIDE_PREFIX)) {
+                String overrideName = name.substring(OVERRIDE_PREFIX.length());
+                props.put(overrideName, substVars(value, name, null, props, callback));
+            } else {
+                props.put(name, substVars(value, name, null, props, callback));
+            }
+        }
+        props.keySet().removeIf(k -> k.startsWith(OVERRIDE_PREFIX));
+    }
+
+    private static void substitute(MavenProperties props, Function<String, String> callback, String name) {
+        String value = props.getProperty(name);
+        if (value == null) {
+            value = callback.apply(name);
+        }
+        if (value != null) {
+            props.put(name, substVars(value, name, null, props, callback));
+        }
+    }
+
+    private static MavenProperties loadPropertiesFile(Path path, boolean failIfNotFound) throws Exception {
+        MavenProperties configProps = new MavenProperties(null, false);
+        try {
+            configProps.load(path);
+        } catch (NoSuchFileException ex) {
+            if (failIfNotFound) {
+                throw ex;
+            }
+        } catch (Exception ex) {
+            System.err.println("Error loading config properties from " + path);
+            System.err.println("Main: " + ex);
+            return configProps;
+        }
+        loadIncludes(INCLUDES_PROPERTY, path, configProps);
+        trimValues(configProps);
+        return configProps;
+    }
+
+    private static void loadIncludes(String propertyName, Path configProp, MavenProperties configProps)
+            throws Exception {
+        String includes = configProps.get(propertyName);
+        if (includes != null) {
+            StringTokenizer st = new StringTokenizer(includes, "?\",", true);
+            if (st.countTokens() > 0) {
+                String location;
+                do {
+                    location = nextLocation(st);
+                    if (location != null) {
+                        boolean mandatory = true;
+                        if (location.startsWith("?")) {
+                            mandatory = false;
+                            location = location.substring(1);
+                        }
+                        Path path = configProp.resolve(location);
```

Comment:
proposition:
```
                        Path path = configProp.getParent().resolve(location);
```

`path` is a path for current properties which contains includes so we should resolve against directory contains properties - not a property file.

it will allow relative path in properties like:
```
${includes}=default.properties
```

Reply:
That was the intent, it was clearly missing a unit test.